File: in_Java/1209 Remove All Adjacent Duplicates in String II.java
class Solution {
    public String removeDuplicates(String s, int k) {
        Stack<String> st = new Stack<>();
        for(char c : s.toCharArray()){
            if(!st.isEmpty() && c == st.peek().charAt(0)){
                String cur = st.peek();
                int curlen = Integer.parseInt(cur.substring(1,cur.length()));
                if(curlen == k-1){
                    while(curlen-- != 0) st.pop();
                }else{
                    st.push(Character.toString(c) + (curlen+1));
                }                
            }else{
                st.push(Character.toString(c) + "1");
            }
        }
        StringBuilder sb = new StringBuilder();
        while(!st.isEmpty()){
            //System.out.println(st.peek());
            sb.insert(0, st.pop().charAt(0));
        }
        return sb.toString();
    }
}

File: in_Python/1209 Remove All Adjacent Duplicates in String II.py
class Solution(object):
    def removeDuplicates(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: str
        """
        st = []
        count = []
        for i in s:
            if len(st) != 0 and st[-1] == i:
                if len(count) != 0:
                    dup = count[-1] + 1
                else:
                    dup = 1
                count.append(dup)
                st.append(i)
            else:
                dup = 1
                st.append(i)
                count.append(dup)
            if dup == k:
                while dup != 0:
                    st.pop()
                    count.pop()
                    dup -= 1
        res = ''
        while len(st) != 0:
            res = st.pop() + res
        return res
            